src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineAsyncTests.cs

```diff
@@ -0,0 +1,340 @@
+        public async Task VerifyFunctionIsAsync()
+        {
+            var helper = new WaitHelper();
+            var func = helper.GetFunction("CustomAsync");
+
+            var config = new PowerFxConfig(null);
+            config.AddFunction(func);
+
+            var engine = new RecalcEngine(config);
+
+            // Can be invoked. 
+            var cts = new CancellationTokenSource();
+
+            var task = engine.EvalAsync("CustomAsync(3)", cts.Token);
+            await Task.Yield();
+
+            // custom func is blocking on our waiter
+            await Task.Delay(TimeSpan.FromMilliseconds(5)); 
+            Assert.False(task.IsCompleted);
+
+            helper.SetResult(15);
+
+            var result = await task;
+
+            Assert.Equal(30.0, result.ToObject());
+        }
+        
+        private static async Task<FormulaValue> WorkerWaitForCancel(FormulaValue[] args, CancellationToken cancel)
+        {
+            // Block forever until cancellation. 
+            await Task.Delay(-1, cancel); // throws TaskCanceledException
+
+            throw new InvalidOperationException($"Shouldn't get here");
+        }
+
+        // Verify cancellation 
+        [Fact]
+        public async Task Cancel()
+        {
+            var func = new CustomAsyncTexlFunction("CustomAsync", DType.Number, DType.Number)
+            {
+                _impl = WorkerWaitForCancel
+            };
+
+            var config = new PowerFxConfig(null);
+            config.AddFunction(func);
+
+            var engine = new RecalcEngine(config);
+
+            // Can be invoked. 
+            var cts = new CancellationTokenSource();
+
+            var task = engine.EvalAsync("CustomAsync(3)", cts.Token);
+            await Task.Yield();
+
+            // custom func is blocking on our waiter
+            await Task.Delay(TimeSpan.FromMilliseconds(5));
+            Assert.False(task.IsCompleted);
+
+            cts.Cancel();
+
+            await Assert.ThrowsAsync<TaskCanceledException>(async () => { await task; });
+        }
+
+        // Verify cancellation 
+        [Fact]
+        public async Task InfiniteLoop()
+        {
+            // Create an expression that will take hang (take very long)
+            var n = 10 * 1000;
+            var expr = $"ForAll(Sequence({n}), ForAll(Sequence({n}), ForAll(Sequence({n}), 5)))";
+
+            var engine = new RecalcEngine();
+
+            // Can be invoked. 
+            var cts = new CancellationTokenSource();
+
+            cts.CancelAfter(5);
+
+            // Eval may never return.             
+            await Assert.ThrowsAsync<OperationCanceledException>(async () =>
+            {
+                await engine.EvalAsync(expr, cts.Token);
+            });
+        }
+
+        // Test interleaved concurrent runs. 
+        // RecalcEngine is single threaded, but the same engine should be able to do multiple evals.
+        [Fact]
+        public async Task ConcurrentInterleave()
```

Comment:
>ConcurrentInterleave [](http://example.com/codeflow?start=26&length=20)

@Felx-B - is this what you had in mind?

Reply:
Yes, it looks like the way we will use it
Thank you